### DISCUSSION TITLE
Representing Data

### DISCUSSION THREAD

**Discussion by slanden** [pdml-lang/pdml-lang.github.io#15] [Ideas]
## Summary
Using PDML for data is not as convenient as it is for markup. In the following sections I describe some ideas for improvement and hope to inspire other ideas or thoughts that I haven't shared below.

## Use Cases
- Data Transfer
- Data Definition
- Data Serialization

Other Affected Features Not (Yet) Covered:
- Defining nodes
- Schemas

## Use Case 1: Data Transfer
The unofficial standard in the Web world is JSON, while XML is more prominant in traditional mobile development. People still use XML despite JSON's ~~existence~~ widespread adoption because they need to describe more complex documents, which hints at a problem with JSON; though less verbose, it's not as robust. I see great potential for PDML as a more lightweight human-readable format to send data around. 

Consider the following JSON data.
```
{"users":[
  {
    "id": 1,
    "name": "a",
    "role": "admin",
    "books": [
      {"id": 1, name: "Cool Book"}
    ],
    "verified": true
  },
  {"id": 2, "name": "b", "role": "basic", "books": [{"id": 1, name: "Cool Book"}], "verified": true},
  {"id": 3, "name": "c", "role": "basic", "books": [{"id": 1, name: "Cool Book"}], "verified": true},
  {"id": 4, "name": "d", "role": "admin", "books": [{"id": 1, name: "Cool Book"}], "verified": true},
  {"id": 5, "name": "e", "role": "basic", "books": [{"id": 1, name: "Cool Book"}], "verified": false},
  {"id": 6, "name": "f", "role": "basic", "books": [{"id": 1, name: "Cool Book"}], "verified": true},
  {"id": 7, "name": "g", "role": "basic", "books": [{"id": 1, name: "Cool Book"}], "verified": true}
]}
```
> I expanded the first item in the list to show the structure, all the other rows have the same.

And here it is without the useless quotes (a plain ol' JavaScript object), which we'll use as the base to compare against:
```
users:[
  {
    id: 1,
    name: "a",
    role: "admin",
    books: [
      {id: 1, name: "Cool Book"}
    ],
    verified: true
  },
  {id: 2, name: 'b', role: 'basic', "books": [{"id": 1, name: "Cool Book"}], verified: true},
  {id: 3, name: 'c', role: 'basic', "books": [{"id": 1, name: "Cool Book"}], verified: true},
  {id: 4, name: 'd', role: 'admin', "books": [{"id": 1, name: "Cool Book"}], verified: true},
  {id: 5, name: 'e', role: 'basic', "books": [{"id": 1, name: "Cool Book"}], verified: false},
  {id: 6, name: 'f', role: 'basic', "books": [{"id": 1, name: "Cool Book"}], verified: true},
  {id: 7, name: 'g', role: 'basic', "books": [{"id": 1, name: "Cool Book"}], verified: true}
]
```

Here's the same in PDML:
```
[users
  [user
    [id 1]
    [name a]
    [role admin]
    [books
      [book [id 1] [name Cool Book]]
    ]
    [verified true]
  ]
  [user [id 2] [name b] [role basic] [books [book [id 1] [name Cool Book]]] [verified true]]
  [user [id 3] [name c] [role basic] [books [book [id 1] [name Cool Book]]] [verified true]]
  [user [id 4] [name d] [role admin] [books [book [id 1] [name Cool Book]]] [verified true]]
  [user [id 5] [name e] [role basic] [books [book [id 1] [name Cool Book]]] [verified false]]
  [user [id 6] [name f] [role basic] [books [book [id 1] [name Cool Book]]] [verified true]]
  [user [id 7] [name g] [role basic] [books [book [id 1] [name Cool Book]]] [verified true]]
]
```

To me, this is slightly more difficult to read (even without the nested `books` node). I think it's that every property is wrapped in `[]`.

> We're not using the "only-attributes" node syntax, and I'll explain why later.

Using the second child item to compare (since the first is formatted) the JavaScript is 91 characters while the PDML is 90. We didn't need the inner commas, trailing commas, or colons, but still ended up about the same.

### Empty Names
We can shave a little off by allowing the child names to be optional.

```
[users
  [
    [id 1]
    [name a]
    [role admin]
    [books
      [book [id 1] [name Cool Book]]
    ]
    [verified true]
  ]
  [[id 2] [name b] [role basic] [books [book [id 1] [name Cool Book]]] [verified true]]
  [[id 3] [name c] [role basic] [books [book [id 1] [name Cool Book]]] [verified true]]
  [[id 4] [name d] [role admin] [books [book [id 1] [name Cool Book]]] [verified true]]
  [[id 5] [name e] [role basic] [books [book [id 1] [name Cool Book]]] [verified false]]
  [[id 6] [name f] [role basic] [books [book [id 1] [name Cool Book]]] [verified true]]
  [[id 7] [name g] [role basic] [books [book [id 1] [name Cool Book]]] [verified true]]
]
```

The reason is, when mapping this data to types in application code, the child name is useless information. It'll be mapped to either a map-thing or an array-thing. In JavaScript for example,

```
let some_users = parse_pdml(fetch_data(https://path-to-users.api));
```
The top level node name `[users]` doesn't get used, because you dump the children into an array-thing called `some_users`. Inside the data, the child names don't matter for the same reason; they get assigned as items in `some_users`. If the data was a map-thing, it would just get parsed like normal *I think*. But, it might make more sense in the next sections.

Anyway, that brings us to 85 characters, 93% of the baseline. From here we can go two different ways.

#### Option A
We use attributes instead. Even though parenthesis are optional if a node only contains attributes, omitting them would mean the parser needs to know that the node should be parsed as attributes. That would require it having the node name in its vocabulary to tell it, "this node is just attributes", or some other special syntax.

By using parenthesis, we save that problem for another section.

```
[users
  [(
    id=1
    name=a
    role=admin
    [books
      [(id=1 name="Cool Book")]
    ]
    verified=true
  )]
  [(id=2 name=b role=basic [books [(id=1 name="Cool Book")]] verified=true)]
  [(id=3 name=c role=basic [books [(id=1 name="Cool Book")]] verified=true)]
  [(id=4 name=d role=admin [books [(id=1 name="Cool Book")]] verified=true)]
  [(id=5 name=e role=basic [books [(id=1 name="Cool Book")]] verified=false)]
  [(id=6 name=f role=basic [books [(id=1 name="Cool Book")]] verified=true)]
  [(id=7 name=g role=basic [books [(id=1 name="Cool Book")]] verified=true)]
]
```

That brings us down to 81% of the baseline and is slightly more readable, in my opinion. Still got that books node though..

As mentioned, the data will be mapped to some type of collection called `users` in the application code. I'm not aware of a good use case for the children not having the same structure; perhaps the closest one is if some fields are missing. In this case the field would need to be something like `field=null`.

#### Option B
Instead, we could allow the parent to tell the parser how to parse each child node. In doing so, we could get close to a delimiter-separated-values list.

```
[users(id=number name=string role=string books=node verified=bool)
  [
    1
    a
    admin
    [books
      [(id=1 name="Cool Book")]
    ]
    true
  ]
  [ 2 b basic [books [(id=1 name="Cool Book")]] true]
  [ 3 c basic [books [(id=1 name="Cool Book")]] true]
  [ 4 d admin [books [(id=1 name="Cool Book")]] true]
  [ 5 e basic [books [(id=1 name="Cool Book")]] false]
  [ 6 f basic [books [(id=1 name="Cool Book")]] true]
  [ 7 g basic [books [(id=1 name="Cool Book")]] true]
]
```

Note that even without a name we need the [PDML separator](https://pdml-lang.dev/docs/basic-specification/#node_separator) (a space in this case) so the first item doesn't become the node's name, which would mean it's missing from the inner content.

We could've used `()` instead of the separator, but the idea here is that maybe these aren't attributes to the parser, but are just content being split on whitespace. I'm not sure if the difference is significant. Then again, that books node means maybe they should be parsed like regular attributes.

So, we cut down quite a bit and are no longer duplicating information. However, we needed to add some up front so the total length of the PDML data is 69% of the JavaScript version. The idea is, for any node that has nested children, you can specify the child's schema in the parent's attributes.

It might also be necessary for the parent's first attribute to explicitly state this, e.g.
`[users (schema id=number name=string role=string books=node verified=bool)]`.

This is so the parser knows to not parse each child's content as text but rather to split by whitespace. That's just a speculation though. Alternatively, we could just wrap in `()` to achieve a similar effect.

Instead of the additional `schema` attribute, the parser could be taught to "parse nodes without a name as attributes". The downsides of this are:
- More specific parsing, and whether the benefit outweighs that cost
- What if it's decided that an empty name should mean something else, like a text* node or for parsers to ignore nodes with empty names, etc.

*Then again, the idea that a node without a name was free text might be a non-problem because text is checked for between nodes, whereas in the case of `[]` or `[ no name]` the parser first saw `[`, which is not free text.

**Potential Issues**

There are a few things I'm concerned about with the approach though.

First, `books=node`. I don't like the idea of using a reserved keyword interspersed with the other attributes that are supposed to resemble the children's schema. That takes the word out of the user's available choices, in this case I chose "node" for demonstration. It also looks like it would resemble a type, but it has a different meaning from its neighboring attributes.

The second thing is although this is valid PDML, I'm not sure about having nodes among attributes or attribute-like things. I'm not certain if this will prohibit or complicate the ideas described, or if it would look confusing or unexpected to people reading their data. Maybe not, I don't know.

The third thing is, this doesn't always mean smaller data. If we applied the same technique to the `books` node, the total size jumps back up to 83% of the JavaScript. That has to do with the fact that `books` is a complex object inside each child, so each child needs to specify the schema for `books`. It'd be better if we could specify it once on the root node..

We might end up needing some sort of header node as the first child, but.. then you need a header node as the first child.

## Use Case 2: Data Definition
I'm not trying to cover writing or defining schemas here, just how they could be affected by our choice of representing data.

We've been using attributes on a node to describe how its children should be parsed. What if we want to describe how the node itself should be parsed? Or, what if we want semantically different attributes from the child-parsing stuff? For example, in schemas..

```
[database library
  [table books
    [column id int unsigned]
    [column name varchar(255)]
    [column author varchar(255)]
    [column published date]

    [index (unique) name author] [- composite key -]
    [index (
      name=published_at_index
      note='When the author published this'
      ) created_at
    ]
    [data(name=string author=string published=date)
      [ "Defining Default Data" "Some Guy" 2023-02-15]
      [ "Useful for Immutable Deployments" "Some Gal" 2023-02-15]
    ]
  ]
]
```

The relevant things to note here are:
- `data` is missing `id` in its attributes, because the database will generate this automatically
- Specifying the types with the column names in data's attributes when their definitions are right above feels unnecessary in this case, as the parser is not mapping these items to application code but passing them to the database language's parser. However, as shown above, the column nodes use SQL syntax while the data node uses a more generic syntax. The reason is, the data node's attribute syntax would have meaning to PDML while the column node's content is just passed to the database being used and uses whichever syntax the database uses.
- the `schema` attribute is missing when we said we *might* need it. I'm not sure about it, especially in the above case.

**Comment by pdml-lang** [comment]
Very interesting ideas, well worth to ponder about.

I will reply in a few days, and suggest a few ideas I had already about the items mentioned in your message: anonymous nodes, positional attributes, types, and schemas.

**Comment by pdml-lang** [comment]
Before further discussions let me first describe a few PDML features related to the subject of your message.
Some of the following features are already implemented, and some might be added in future PDML versions.
These features allow to customize PDML and use alternative/shorter syntaxes for specific nodes.

## Anonymous Nodes

Sometimes the name of a node isn't relevant.
This begs the question: Should we support anonymous nodes (nodes that have no name)?

You mentioned a typical example (list elements) where the node name isn't needed:

```
[users
    [user ...]
    [user ...]
    [user ...]
]
```

In the above example, the name `user` is redundant and could be omitted in all children of `users`.

A first obvious approach would be to simply leave off the node name and write:

```
[users
    [ ...]
    [ ...]
    [ ...]
]
```

Note the space after the `[`, which is required.

This solution has two drawbacks:

1. Using a space to declare an anonymous node is error-prone.
    Generally speaking, using whitespace to define structure is most often a bad idea.
    Spaces are not always well visible, especially when non-monospace fonts are used.
    Moreover, one must be careful to interpret the code correctly, and the semantics change if someone (accidentally) removes or adds a space:
    ```
    [foo bar]     // node 'foo' with content 'bar'
    [ foo bar]    // anonymous node with content 'foo bar'
    [  foo bar]   // anonymous node with content ' foo bar'

    [foo]         // empty node 'foo'
    [ foo]        // anonymous node with content 'foo'
    [  foo]       // anonymous node with content ' foo'

    [ ]           // anonymous empty node
    [  ]          // anonymous node with one space as content
    []            // invalid
    ```

2. Using a space after the `[` is invalid in _Core PDML_.
    Hence, we would need to introduce an additional _extension_ to support anonymous nodes.
    However, documents with anonymous nodes could not be parsed by a parser that doesn't support this extension.

To avoid these drawbacks I suggest using `_` as an explicit symbol for anonymous nodes.
Hence, the above example would be written as follows:
```
[users
    [_ ...]
    [_ ...]
    [_ ...]
]
```

Obviously this is not as succinct as just using a space, but there are considerable advantages:

- According to the Core PDML [naming rules](https://pdml-lang.dev/docs/core/specification/index.html#node_name), `_` is a valid name.
  Hence, the document can still be parsed by a parser that supports only _Core PDML_.
  There is no need to introduce a dedicated extension for anonymous nodes.
  We could just introduce the _convention_ (or maybe even the _rule_) that `_` is reserved for anonymous nodes.

- The presence of the `_` clearly indicates the author's intent to use an _anonymous_ node.
  There is no risk of deleting a space and thereby changing the semantics.

- The `_` is already used in some programming languages to denote unused variables (e.g. in pattern matching, where a member is not used).
  Thus, we would use a syntax that is already familiar to some people.

Current development state: Anonymous nodes are currently not specified in PDML.


## Positional Attributes

PDML currently only supports _named_ attributes, e.g.

```
[user (id=1 name=a role=admin verified=true)]
```

_Positional_ attributes would allow to leave off the attribute names, and just write:

```
[user (1 a admin true)]
```

However, this obviously much shorter syntax comes at a price:

- Positional attributes potentially decrease readability.
    For example:
    ```
    [panel (200 100)
        ...
    ]
    ```
    is less readable (especially for new users) than:
    ```
    [panel (width=200 height=100)
        ...
    ]
    ```

- Positional attributes are error-prone, because the order matters.
  For example, if the order of attributes changes, then all documents must be updated, otherwise there is a risk of silently reading wrong data.
  Named attributes don't have this drawback.

- Adding positional attributes increases complexity and requires new PDML rules to be introduced, especially if we allow named and positional attributes to be mixed.

- Parsing attributes becomes more challenging.

- Developing third-party tools (e.g. editor plugins) becomes more challenging too, and for some plugins it might even not be possible to support named and positional attributes (depends on the plugin architecture and limits).

We should therefore think twice and not rush adding positional attributes.
If the community asks and insists for them, they should probably be disabled by default, and enabled only if a dedicated parser configuration flag is explicitly set to true.

Current development state: Positional attributes are supported in [pp-libs](https://github.com/pdml-lang/pp-libs) (the base library used by PDML), but not used in PDML.


## Types

Note for readers new to PDML types: Please read first: [Types](https://pdml-lang.dev/docs/extensions/user_manual/index.html#type_nodes)

Types enable PDML documents to use _any_ syntax for the content of specific nodes.

In your message you provided the following example code that uses the standard PDML syntax:

```
[user [id 2] [name b] [role basic] [books [book [id 1] [name Cool Book]]] [verified true]]
```

Types allow us to define a type `book` that enables us, for example, to use the following syntax for books:

```
[book 1, "Cool Book"]
```

We could also define a type `user` that changes the syntax for user nodes to:

```
[user 2, b, basic, true [books [book 1, "Cool Book"]]]
```

If we have a good reason to support JSON/XML/YAML syntax, or any DSL, we can do that too.
For example we could define a specific type to use standard JSON syntax (inside a PDML document) for `user` nodes:

```
[users
    [user {
        "id": 1,
        "name": "a",
        "role": "admin",
        "books": [
            {"id": 1, "name": "Cool Book"}
        ],
        "verified": true
    }]
    [user ...]
]
```

Current development state:

Types are implemented as an experimental feature in PDML.

In PML, for example, type type `raw_text` is used for nodes that contain plain text which is not parsed as PDML code. Example:

```
    [code
        ~~~
        int i = 1;
        ~~~
    ]
```

PDML currently only supports a very limited set of predefined types.
More standard types will be added in future PDML versions (easy to do).
Specific domain types can currently only be added in Java applications that use the PDML Java libraries as a dependency (.jar files), and programmatically configure the PDML parser.
A future PDML version should enable users to add types in external configuration files that can be shared.


## Node Specifications

A _node specification_ (node spec) is used to define properties of one or several PDML nodes.

The following properties can be specified (all optional):

- The node's attributes. For each attribute you can define:
    - Name
    - Optional alternative names
    - Type (including parser)
    - Validator
    - Default value
    - Documentation

- The node's type (including content parser)
- A validator for the node content
- Documentation
- Other properties

A list of node specs can be passed to the PDML parser configuration object.
Right after parsing a node name in a PDML document, the parser looks for a node spec, and if a spec is found then it is applied to the attributes and content of the node being parsed.

Current development state: [Implemented](https://github.com/pdml-lang/full-pdml-impl/tree/main/pdml-parser/src/main/java/dev/pdml/parser/nodespec) already and used in PML.


## Schemas

Schemas are still on the to-do list, but ...

Schemas would allow us to:
- Define types
- Associate types to nodes
- Validate nodes and whole PDML documents

A schema is itself a PDML document that specifies the structure of target PDML data documents.

For example, a schema would allow us to specify that:
- The root node must contain exactly one node named `users`.
- Node `users` can only contain nodes with name `user`.
- Node `user` is a record with fields `id`, `name`, etc., and must be a child of `users`.

A schema can be defined internally (in the PDML document that contains the data) or externally (e.g. an external file or URL that is referenced in the data document).

Here is a quick-and-dirty example, just to show how a PDML document containing the schema and data could look like:

```
[schema
    [types
        [type
            [name users]
            [kind list<user>]
        ]
        [type
            [name user]
            [kind record]
            [fields
                [field
                    [id id]
                    [type number]
                    [unique yes]
                ]
                [field
                    [id name]
                    [type string]
                ]
                [field ...]
                [field ...]
            ]
        ]
        [type
            [name books]
            ...
        ]
        [type
            [name book]
            ...
        ]
    ]
    [nodes
        [node
            [name users]
            [type users]
        ]
        [node
            [name user]
            [type user]
            [check
                ~~~
                // Using Java code to write
                // validation code !!!
                context.checkParent ( "users" )
                ~~~
            ]
        ]
        ...
    ]
]
[data
    [users
        [user ...]
        [user ...]
    ]
]
```

Current development state: PDML schemas aren't yet specified/implemented.

Note: Schemas (combined with the option to use Java as a scripting language for node validation and specific parsing) will make PDML much more powerful and versatile.
Schemas are therefore a top-priority and an experimental (but well thought-out) version should be implemented in the near future.
I have already started a side project to integrate Java scripting (not Javascript!) in an efficient way, and then use it inside of schemas and PDML script nodes.

**Reply by slanden** [reply]
**Anonymous Nodes**
> Spaces are not always well visible, especially when non-monospace fonts are used.

This was my first thought, and you made another good point about when extra spaces are involved.

> `_` as an explicit symbol for anonymous nodes

I like this because, as you said, it's used in some other languages to denote unused variables. That's familiar to me and felt natural.

**Positional Attributes**
> Positional attributes potentially decrease readability

True; I think if something like this was added, there needs to be an explicit statement/syntax before and near its use basically saying, "attention, these attributes have a specified order".

I agree that considerable thought should go into this, maybe we can find a better way to cut down on noise in the situations this would be useful.

**Types**
> `[book 1, "Cool Book"]`

This looks and sounds like my original use case for positional arguments when I talked about *Empty Names > Option B*:
> Instead, we could allow the parent to tell the parser how to parse each child node. In doing so, we could get close to a delimiter-separated-values list.

To use for data transfer though, you'd need to define the type somewhere at the top level so the receiving parser would know what to do (unless it's a built in type of course).

However, You couldn't combine that with the use of anonymous nodes, i.e. `[_ 1, "Cool Book"]`, because it would apply to all anonymous nodes and may not be what the author wanted.

**Node Specifications**
Whether this would be a good fit or not, depends.. When I think of node specifications, I think they're how you would define PML and other domain-specific languages, which is more than what I'm talking about here.

Yeah, it depends; if you need to pass the node spec with the data, it depends on how much extra that is. If you have a Real-time financial application pulling in transaction info, you wouldn't want to send a large schema with every response. But, if it was just as minimal as the data items being sent, then maybe.

You could do some content negotiation beforehand I suppose, but I think enabling a more stateless response would be a better foundation.

I've been thinking about the ability to define nodes too though, I'll write up some thoughts in a separate discussion. Probably nothing you haven't been thinking about already though.

**Schemas**
Schemas in the way you described, is a separate feature altogether I think. What I'm talking about is syntax for terse data representation the receiving end could map to application data, say, a JavaScript object. This is opposed to validating that the structure meets business rules like, "max length" or "one of these values".

**Comment by pml-lang** [comment]
> we could allow the parent to tell the parser how to parse each child node.

This is an interesting idea to be explored, in addition to types, node specifications, and schemas.

> When I think of node specifications, I think they're how you would define PML and other domain-specific languages.

True.

> I've been thinking about the ability to define nodes too though, I'll write up some thoughts in a separate discussion.

Great! We can then discuss and compare several approaches.
